src/pybind/mgr/nfs/export.py

```diff
@@ -658,6 +660,10 @@ def create_cephfs_export(self,
                              access_type: str,
                              clients: list = [],
                              sectype: Optional[List[str]] = None) -> Dict[str, Any]:
+
+        path_exists = check_cephfs_path(self.fs_client, fs_name, path)
+        if path_exists[0] != 0:
+            raise NFSException(f"Failed to create export: {path_exists[1]}")
```

Comment:
I think it would be cleaner and more pythonic to raise the exception directly from `check_cephfs_path`. One of my goals of adding the object formatter decorators is to allow a more idomatic python style in most mgr code.

Reply:
> > @phlogistonjohn Hi there. Seeing your contributions to the mgr/nfs code, pinging you case you have some insights on the issue this PR is facing.
> > **Summary of the code before moving to the issue: This PR uses `mgr_util.py`'s `CephfsClient` and `open_filesystem()` in `export.py` to open a fs handle to check if the cephfs path is valid or not before creating cephfs exports.**
> > Issues are not faced in teuthology, it's all green but rather make check tests never complete, runs till timeout i.e. 3 hours. I had to run tox tests locally to find out the actual issue that at the time of exit, the run would halt and need manual interruption (CTRL + C) else it would wait till infinity.
> > At first the issue seemed to be with `fs` not monkey patched in `test_nfs.py` and thus leaded to Exception getting caught and raised which I fixed by wrapping the code that checks for cephfs path into a helper function under `utils.py` and mocking it as true in `test_nfs.py` ([52c5812](https://github.com/ceph/ceph/commit/52c58123aad2cd1ef5e67704b559fcd7ec0e851a)). Once this got sorted the tests passed but the halting issue still persisted, then i pushed a commit [a9cb9e5](https://github.com/ceph/ceph/commit/a9cb9e5f36afa0bba0ab6deedbd00dda93e5ad0c) that comments a test case(how did i figure this out? trial and error with every test case) in `test_nfs.py` and also monkey patches `CephfsClient` but very vaguely, i mean we would need all the things to be patched but this was just to see if bypassing it makes it work or not and to my surprise it went fine and make check passed. So there's a lot weird going on. @vshankar and I agree to completely monkey patch `CephfsClient` which should this issue however we thought to ping and ask you if you have any thoughts on this. Especially the mystery of the test case that is commented out.
> 
> I haven't dug into the behavior of CephfsClient much myself but it seems that if adding:
> 
> ```
>                 mock.patch('nfs.export.check_cephfs_path',
>                            return_value=(0, "cephfs path is valid")), \
> ```
> 
> did not fix the hang but adding:
> 
> ```
>                 mock.patch('nfs.export.CephfsClient', return_value=mock.MagicMock()), \
> ```
> 
> did I bet the fact that the CephfsClient is created and assigned in ExportMgr.**init** is an issue:
> 

You're right, it's the instantiation of `CephfsClient` that is breaking the things
> ```
> self.fs_client = CephfsClient(mgr)
> ```
> 
> IOW, if there's state that needs to connect to some "real" ceph process this could interfere with the unit tests.
> 
> I really like that you moved the path check to a function anyway, I think that is a good change regardless of the test hang. But I do think it would be better to only mock `check_cephfs_path` if possible. Could you experiment with moving the instantiation of the CephfsClient into check_cephfs_path instead of in `__init__`?
Well this was the initial plan and I wish it was feasible but having helper instantiate `CephfsClient` would mean a new fs handle at every cephfs export creation and which doesn't sound best approach.
> 
> I'm not sure about the other aspect. I would probably need to experiment with it locally myself. If you think that would help please just say so, otherwise I'll let you try the above first.
> 
If you can experiment locally, it would be awesome to have your perspective, just FYI make sure you don't use the last commit, its some experimental change that i pushed to bypass CephfsClient instantiation and have a test case commented(this test case once commented + CephfsClient bypassed does lead to make check success which I can't understand)
> Once you have things working I'll be happy to review this as well. I will make one small style suggestion today.

Reply:
> I think it would be cleaner and more pythonic to raise the exception directly from `check_cephfs_path`. One of my goals of adding the object formatter decorators is to allow a more idomatic python style in most mgr code.

ACK

Reply:
So I ran the pr locally and can confirm the pause and that moving the CephfsClient instantiation out of `__init__` works around that. I have no idea why test_export_validate is involved though.

I ran gdb to see if I could figure out the reason for the pause. Unfortunately, all I could tell is that the main thread is stuck on a lock in the python threading module:
```
Thread 1 (Thread 0x7f834c1b1740 (LWP 39495) "pytest"):
Traceback (most recent call first):
  File "/usr/lib64/python3.11/threading.py", line 1574, in _shutdown
    lock.acquire()
```
But I couldn't learn much more from that.

I made my own little hacky workaround by changing:
```python
def check_cephfs_path(client: CephfsClient, fs: str, path: str) -> Tuple[int, str]:
```
to
```python
def check_cephfs_path(mgr: 'Module', fs: str, path: str) -> Tuple[int, str]:
```
and doing the CephfsClient instantiation inside check_cephfs_path.
I don't know if there's a performance impact for that. I did not run it in a real mgr.

But I figure if there is a performance difference you could do something like add:
```python
@functools.cache    
def _get_cephfs_client(mgr: 'Module') -> CephfsClient:
    return CephfsClient(mgr)
                        
```

to have a memoized fetcher that is called from inside check_cephfs_path.  I dunno if that appeals to you.

Reply:
A portion of my reply got into your reply, it was for:

> I really like that you moved the path check to a function anyway, I think that is a good change regardless of the test hang. But I do think it would be better to only mock check_cephfs_path if possible. Could you experiment with moving the instantiation of the CephfsClient into check_cephfs_path instead of in __init__?


Well this was the initial plan and I wish it was feasible but having helper instantiate CephfsClient would mean a new fs handle at every cephfs export creation and which doesn't sound best approach.

For your recent response, I really like the idea of 
```
@functools.cache    
def _get_cephfs_client(mgr: 'Module') -> CephfsClient:
    return CephfsClient(mgr)
 ```

I tried this locally if it can bailout the main python thread(i know you were talking about instantiation from within the helper but again I wanted to give it a try) but it still halted. Mocking this helper might be doable but I have to mock every member of the class.

I was also thinking of creating a singleton class in `test_nfs.py` that inherits `CephfsClient` that is thread safe, so that we don't have many instances to deal with which might in turn help us resolve this?

Reply:
OK. @dparmar18 I thought more about the singleton object you mentioned yesterday. Do we plan to instantiate it in mgr_util.py? That looks _a bit_ off-putting to me. Can we store the single object (reference) somewhere else? (maybe in the MgrModule class)?

Reply:
mgr_util?, i think you misinterpreted it, it was `utilts.py`. we would instantiate it in the helper

Reply:
ah ok. I heard it has mgr_util.py is why I was a bit concerned. I'm fine with having it instantiated in utils.py provided its a singleton.